src/utils/utils_log.c

```diff
@@ -0,0 +1,252 @@
+    case LOG_WARNING:
+        return "WARN";
+    default:
+        ASSERT(0);
+        return "";
+    }
+}
+
+static const char *match_env_var(const char *envVar, const char *prefix,
+                                 const char **extraArg) {
+    const char *found = strstr(envVar, prefix);
+    // ensure that found string is first on list or it's a separating semicolon
+    if (!found) {
+        return NULL;
+    }
+    // if found string is not at the beginning of envVar ensure it's preceded by ';'
+    if (found != envVar && *(found - 1) != ';') {
+        return NULL;
+    }
+
+    const char *endFound = found + strlen(prefix);
+    if (!extraArg) {
+        // if there is no argument, matched string should end with ';' or '\0'
+        if (*endFound != '\0' && *endFound != ';') {
+            return NULL;
+        }
+        return found;
+    }
+
+    // check if matched string ends with '='
+    if (*endFound != '=') {
+        return NULL;
+    }
+
+    *extraArg = endFound + 1;
+
+    return found;
+}
+
+void utils_log(utils_log_level_t level, const char *format, ...) {
+
+    if (!loggerConfig.output) {
+        return; //logger not enabled
+    }
+    if (level < loggerConfig.level) {
+        return;
+    }
+
+#if defined(_WIN32)
+    DWORD pid = GetCurrentProcessId();
+    DWORD tid = GetCurrentThreadId();
+#elif defined(__APPLE__)
+    pid_t pid = getpid();
+    uint64_t tid64;
+    pthread_threadid_np(NULL, &tid64);
+    pid_t tid = (pid_t)tid64;
+#else
+    pid_t pid = getpid();
+    pid_t tid = gettid();
+#endif
+
+    char buffer[LOG_MAX];
+    va_list args;
+    va_start(args, format);
+    int ret = vsnprintf(buffer, sizeof(buffer), format, args);
+    const char *overflow = "";
+    if (ret >= (intptr_t)sizeof(buffer)) {
+        //TODO: alloc bigger buffer with base alloc
+        overflow = "[truncated...]";
+    }
+    va_end(args);
+    char header[LOG_HEADER];
+    char *h_pos = header;
+    memset(header, 0, sizeof(header));
+    *h_pos++ = '[';
+
+    if (loggerConfig.timestamp) {
+        time_t now = time(NULL);
+        struct tm tm_info;
+#ifdef _WIN32
+        localtime_s(&tm_info, &now);
+#else
+        localtime_r(&now, &tm_info);
+#endif
+
+        ASSERT((intptr_t)sizeof(header) > (h_pos - header));
+        h_pos += strftime(h_pos, sizeof(header) - (h_pos - header),
+                          "%Y-%m-%dT%H:%M:%S ", &tm_info);
+    }
+
+    if (loggerConfig.pid) {
+        ASSERT((intptr_t)sizeof(header) > (h_pos - header));
+        h_pos += snprintf(h_pos, sizeof(header) - (h_pos - header),
+                          "PID:%-6lu TID:%-6lu ", (unsigned long)pid,
+                          (unsigned long)tid);
+    }
+
+    // We take twice header size here to ensure that
+    // we have space for log level and overflow string
+    // otherwise -Wformat-truncation might be thrown by compiler
+    char logLine[LOG_MAX + LOG_HEADER * 2];
+    snprintf(logLine, sizeof(logLine), "%s%-5s UMF] %s%s\n", header,
```

Comment:
nit: `[` could be extracted from the header and placed directly here, easier to read.

Reply:
Done